src/main/java/org/apache/maven/plugins/toolchain/jdk/ToolchainDiscoverer.java

```diff
@@ -0,0 +1,433 @@
+    public static final String RUNTIME_VERSION = "runtime.version";
+    public static final String VENDOR = "vendor";
+    public static final String VENDOR_VERSION = "vendor.version";
+    public static final String[] PROPERTIES = {VERSION, RUNTIME_NAME, RUNTIME_VERSION, VENDOR, VENDOR_VERSION};
+
+    public static final String DISCOVERED_TOOLCHAINS_CACHE_XML = ".m2/discovered-toolchains-cache.xml";
+
+    public static final String CURRENT = "current";
+    public static final String ENV = "env";
+    public static final String LTS = "lts";
+
+    public static final String JDK_HOME = "jdkHome";
+    public static final String JAVA_HOME = "java.home";
+
+    private static final String COMMA = ",";
+    public static final String USER_HOME = "user.home";
+
+    private final Logger log = LoggerFactory.getLogger(getClass());
+
+    private Map<Path, ToolchainModel> cache;
+    private boolean cacheModified;
+
+    /**
+     * Build the model for the current JDK toolchain
+     */
+    public Optional<ToolchainModel> getCurrentJdkToolchain() {
+        Path currentJdkHome = getCanonicalPath(Paths.get(System.getProperty(JAVA_HOME)));
+        if (hasJavaC(currentJdkHome)) {
+            // in case the current JVM is not a JDK
+            return Optional.empty();
+        }
+        ToolchainModel model = new ToolchainModel();
+        model.setType(TOOLCHAIN_TYPE_JDK);
+        Stream.of(PROPERTIES).forEach(k -> {
+            String v = System.getProperty(JAVA + k);
+            if (v != null) {
+                model.addProvide(k.substring(JAVA.length()), v);
+            }
+        });
+        model.addProvide(CURRENT, "true");
+        Xpp3Dom config = new Xpp3Dom("configuration");
+        Xpp3Dom jdkHome = new Xpp3Dom(JDK_HOME);
+        jdkHome.setValue(currentJdkHome.toString());
+        config.addChild(jdkHome);
+        model.setConfiguration(config);
+        return Optional.of(model);
+    }
+
+    public PersistedToolchains discoverToolchains() {
+        return discoverToolchains(LTS + COMMA + VERSION + COMMA + VENDOR);
+    }
+
+    /**
+     * Returns a PersistedToolchains object containing a list of discovered toolchains,
+     * never <code>null</code>.
+     */
+    public PersistedToolchains discoverToolchains(String comparator) {
+        try {
+            Set<Path> jdks = findJdks();
+            log.info("Found " + jdks.size() + " possible jdks: " + jdks);
+            cacheModified = false;
+            readCache();
+            Map<Path, Map<String, String>> flags = new HashMap<>();
+            Path currentJdkHome = getCanonicalPath(Paths.get(System.getProperty(JAVA_HOME)));
+            flags.computeIfAbsent(currentJdkHome, p -> new HashMap<>()).put(CURRENT, "true");
+            // check environment variables for JAVA{xx}_HOME
+            System.getenv().entrySet().stream()
+                    .filter(e -> e.getKey().startsWith("JAVA") && e.getKey().endsWith("_HOME"))
```

Comment:
maybe a regex including the digits?

Reply:
Not sure why, this test also captures `JAVA_HOME` which is fine.  I'm not sure we need to restrict to using numbers, even if that's what people usually do.  The only place it's used is when matching against a user supplied value in the `env` filter during selection, but it could be anything, especially something like: `JAVA_GRAAL_17_HOME` or whatever.